File: decoder/decode.py
#!/usr/bin/env python3
import base64
import hashlib
import json
import logging
import os
import threading
import time
from datetime import datetime
from urllib.parse import urlparse

import paho.mqtt.client as mqtt
from meshcoredecoder import MeshCoreDecoder
from meshcoredecoder.crypto import MeshCoreKeyStore
from meshcoredecoder.types.crypto import DecryptionOptions
from meshcoredecoder.types.enums import PayloadType

logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s %(levelname)s %(message)s',
    handlers=[logging.StreamHandler()]
)
log = logging.getLogger(__name__)

# --- Config ---
INPUT_BROKER      = os.environ.get('INPUT_BROKER',      'mqtt://localhost:1884')
OUTPUT_BROKER     = os.environ.get('OUTPUT_BROKER',     'mqtt://192.168.11.10:1883')
INPUT_USER        = os.environ.get('INPUT_MQTT_USER')
INPUT_PASS        = os.environ.get('INPUT_MQTT_PASS')
OUTPUT_USER       = os.environ.get('OUTPUT_MQTT_USER',  'mqtt')
OUTPUT_PASS       = os.environ.get('OUTPUT_MQTT_PASS',  'mqtt')
PUBLIC_SECRET_B64 = os.environ.get('PUBLIC_SECRET',     'izOH6cXN6mrJ5e26oRXNcg==')
HASHTAGS          = [t.strip() for t in os.environ.get('HASHTAGS', 'test,northwest').split(',') if t.strip()]

# --- Key store ---
public_secret_hex = base64.b64decode(PUBLIC_SECRET_B64).hex()
channel_secrets = [
    public_secret_hex,
    *[hashlib.sha256('#{}'.format(tag).encode()).hexdigest()[:32] for tag in HASHTAGS],
]
key_store = MeshCoreKeyStore({'channel_secrets': channel_secrets})
decode_options = DecryptionOptions(key_store=key_store)


PAYLOAD_TYPE_NAMES = {
    PayloadType.Request:     'request',
    PayloadType.Response:    'response',
    PayloadType.TextMessage: 'text_message',
    PayloadType.Ack:         'ack',
    PayloadType.Advert:      'advert',
    PayloadType.GroupText:   'group_text',
    PayloadType.GroupData:   'group_data',
    PayloadType.AnonRequest: 'anon_request',
    PayloadType.Path:        'path',
    PayloadType.Trace:       'trace',
    PayloadType.Multipart:   'multipart',
    PayloadType.Control:     'control',
    PayloadType.RawCustom:   'raw_custom',
}


def channel_hash_from_secret(secret_hex):
    return hashlib.sha256(bytes.fromhex(secret_hex)).hexdigest()[:2].upper()


channel_names = {
    channel_hash_from_secret(public_secret_hex): 'Public',
    **{
        channel_hash_from_secret(
            hashlib.sha256('#{}'.format(tag).encode()).hexdigest()[:32]
        ): '#{}'.format(tag)
        for tag in HASHTAGS
    },
}

# --- Dedup cache ---
EXPIRY_S = 10 * 60
_seen = {}
_seen_lock = threading.Lock()


def has_seen(h):
    with _seen_lock:
        ts = _seen.get(h)
        if ts is None:
            return False
        if time.monotonic() - ts > EXPIRY_S:
            del _seen[h]
            return False
        return True


def add_seen(h):
    with _seen_lock:
        _seen[h] = time.monotonic()


def _cleanup_seen():
    while True:
        time.sleep(EXPIRY_S)
        cutoff = time.monotonic() - EXPIRY_S
        with _seen_lock:
            expired = [k for k, v in _seen.items() if v < cutoff]
            for k in expired:
                del _seen[k]


threading.Thread(target=_cleanup_seen, daemon=True).start()


# --- Helpers ---
def parse_broker(url):
    p = urlparse(url)
    return p.hostname, p.port or 1883


def to_unix(timestamp_str):
    if not timestamp_str:
        return None
    try:
        return int(datetime.fromisoformat(timestamp_str.replace('Z', '+00:00')).timestamp())
    except Exception:
        return None



# --- Output MQTT ---
out_host, out_port = parse_broker(OUTPUT_BROKER)
output_client = mqtt.Client(mqtt.CallbackAPIVersion.VERSION2, client_id='meshcore-decoder-out')
if OUTPUT_USER:
    output_client.username_pw_set(OUTPUT_USER, OUTPUT_PASS)


def on_output_connect(client, userdata, connect_flags, reason_code, properties):
    if reason_code.is_failure:
        log.error('Output connection failed: %s', reason_code)
    else:
        log.info('Output connected: %s', OUTPUT_BROKER)


def on_output_disconnect(client, userdata, disconnect_flags, reason_code, properties):
    if reason_code.is_failure:
        log.warning('Output disconnected unexpectedly: %s', reason_code)


output_client.on_connect = on_output_connect
output_client.on_disconnect = on_output_disconnect
output_client.connect(out_host, out_port)
output_client.loop_start()


def publish(type_name, payload):
    output_client.publish('meshcore/{}'.format(type_name), json.dumps(payload, default=str), qos=0)


def publish_advert(payload):
    msg = json.dumps(payload, default=str)
    output_client.publish('meshcore/advert', msg, qos=0)
    output_client.publish('meshcore/advert/{}'.format(payload['publicKey']), msg, qos=0, retain=True)


# --- Message handler ---
def on_message(client, userdata, msg):
    if not msg.topic.endswith('/packets'):
        return
    try:
        data = json.loads(msg.payload.decode())
        if not data.get('raw') or not data.get('hash'):
            return
        if has_seen(data['hash']):
            return
        add_seen(data['hash'])

        packet = MeshCoreDecoder.decode(data['raw'], decode_options)
        if packet is None:
            return

        path = list(getattr(packet, 'path', None) or [])
        hops = len(path)
        path_hash_bytes = len(path[0]) // 2 if path else None
        received_unix = to_unix(data.get('timestamp'))

        payload_type = getattr(packet, 'payload_type', None)
        type_name = PAYLOAD_TYPE_NAMES.get(payload_type, 'unknown_{}'.format(payload_type))

        def to_float(v):
            try:
                return float(v) if v is not None else None
            except (ValueError, TypeError):
                return None

        common = {
            'type':              type_name,
            'receivedTimestamp': data.get('timestamp'),
            'receivedUnix':      received_unix,
            'packetHash':        data['hash'],
            'path':              path,
            'hops':              hops,
            'pathHashBytes':     path_hash_bytes,
            'snr':               to_float(data.get('SNR')),
            'rssi':              to_float(data.get('RSSI')),
            'score':             to_float(data.get('score')),
        }

        if isinstance(packet.payload, dict):
            decoded = packet.payload.get('decoded')
        else:
            decoded = getattr(packet.payload, 'decoded', None)

        # Advert
        if payload_type == PayloadType.Advert:
            app_data = getattr(decoded, 'app_data', {}) or {}
            location = app_data.get('location')
            device_role = app_data.get('device_role')
            public_key = getattr(decoded, 'public_key', None)
            signature_valid = getattr(decoded, 'signature_valid', None)
            out = dict(common)
            if public_key is not None:
                out['publicKey'] = public_key
            if signature_valid is not None:
                out['signatureValid'] = signature_valid
            name = app_data.get('name')
            if name:
                out['name'] = name
            if location:
                out['lat'] = location.get('latitude')
                out['lon'] = location.get('longitude')
            if device_role is not None:
                role_value = device_role.value if hasattr(device_role, 'value') else device_role
                role_name = device_role.name if hasattr(device_role, 'name') else str(device_role)
                out['deviceRole'] = role_value
                out['deviceRoleName'] = role_name
            if out.get('publicKey'):
                publish_advert(out)
            else:
                publish('advert', out)
            return

        # Group text
        if payload_type == PayloadType.GroupText:
            channel_hash = getattr(decoded, 'channel_hash', None)
            if channel_hash:
                channel_hash = channel_hash.upper()
            channel = channel_names.get(channel_hash, 'unknown({})'.format(channel_hash))
            decrypted = getattr(decoded, 'decrypted', None)
            decrypted_data = decrypted or {}
            sender    = decrypted_data.get('sender')
            message   = decrypted_data.get('message')
            sent_unix = decrypted_data.get('timestamp')
            propagation = (received_unix - sent_unix) if (received_unix and sent_unix) else None
            out = dict(common)
            out.update({
                'channel':         channel,
                'channelHash':     channel_hash,
                'decrypted':       decrypted is not None,
                'sentUnix':        sent_unix,
                'propagationSecs': propagation,
            })
            if sender is not None:
                out['sender'] = sender
            if message is not None:
                out['message'] = message
            publish('group_text', out)
            return

        # Encrypted direct message types we can't decrypt
        ENCRYPTED_TYPES = {
            PayloadType.TextMessage, PayloadType.Request,
            PayloadType.Response, PayloadType.AnonRequest,
        }
        if payload_type in ENCRYPTED_TYPES:
            publish(type_name, dict(common, decrypted=False))
            return

        # Everything else
        publish(type_name, common)

    except Exception as e:
        log.error('Decode error: %s', e)


# --- Input MQTT ---
in_host, in_port = parse_broker(INPUT_BROKER)
input_client = mqtt.Client(mqtt.CallbackAPIVersion.VERSION2, client_id='meshcore-decoder-in')
if INPUT_USER:
    input_client.username_pw_set(INPUT_USER, INPUT_PASS)


def on_input_connect(client, userdata, connect_flags, reason_code, properties):
    if reason_code.is_failure:
        log.error('Input connection failed: %s', reason_code)
    else:
        log.info('Input connected: %s', INPUT_BROKER)
        client.subscribe('meshcore/#')
        log.info('Subscribed to meshcore/#')


def on_input_disconnect(client, userdata, disconnect_flags, reason_code, properties):
    if reason_code.is_failure:
        log.warning('Input disconnected unexpectedly: %s', reason_code)


input_client.on_connect = on_input_connect
input_client.on_disconnect = on_input_disconnect
input_client.on_message = on_message
input_client.connect(in_host, in_port)
input_client.loop_forever()

File: decode.js
const { MeshCoreDecoder } = require('@michaelhart/meshcore-decoder');
const crypto = require('crypto');
const mqtt = require('mqtt');

// --- Config from environment ---
const INPUT_BROKER  = process.env.INPUT_BROKER  || 'mqtt://localhost:1883';
const OUTPUT_BROKER = process.env.OUTPUT_BROKER || 'mqtt://192.168.11.10:1883';
const INPUT_USER    = process.env.INPUT_MQTT_USER  || undefined;
const INPUT_PASS    = process.env.INPUT_MQTT_PASS  || undefined;
const OUTPUT_USER   = process.env.OUTPUT_MQTT_USER || 'mqtt';
const OUTPUT_PASS   = process.env.OUTPUT_MQTT_PASS || 'mqtt';
const PUBLIC_SECRET = process.env.PUBLIC_SECRET   || 'izOH6cXN6mrJ5e26oRXNcg==';
const HASHTAGS      = (process.env.HASHTAGS || 'test,northwest').split(',').map(s => s.trim()).filter(Boolean);

// --- Key store ---
const publicSecretHex = Buffer.from(PUBLIC_SECRET, 'base64').toString('hex');
const channelSecrets = [
	publicSecretHex,
	...HASHTAGS.map(tag => crypto.createHash('sha256').update('#' + tag).digest('hex').slice(0, 32))
];
const keyStore = MeshCoreDecoder.createKeyStore({ channelSecrets });

const PAYLOAD_TYPES = {
	0x00: 'request',
	0x01: 'response',
	0x02: 'text_message',
	0x03: 'ack',
	0x04: 'advert',
	0x05: 'group_text',
	0x06: 'group_data',
	0x07: 'anon_request',
	0x08: 'returned_path',
	0x09: 'trace',
	0x0A: 'multipart',
	0x0B: 'control',
	0x0F: 'raw_custom'
};

const DEVICE_ROLES = {
	0: 'client',
	1: 'client_mute',
	2: 'repeater',
	3: 'room_server',
	4: 'sensor',
	5: 'kiss_modem'
};

function channelHashFromSecret(secretHex) {
	return crypto.createHash('sha256').update(Buffer.from(secretHex, 'hex')).digest('hex').slice(0, 2).toUpperCase();
}

const channelNames = {
	[channelHashFromSecret(publicSecretHex)]: 'Public',
	...Object.fromEntries(
		HASHTAGS.map(tag => {
			const secret = crypto.createHash('sha256').update('#' + tag).digest('hex').slice(0, 32);
			return [channelHashFromSecret(secret), '#' + tag];
		})
	)
};

// --- Dedup cache ---
const EXPIRY_MS = 10 * 60 * 1000;
const seen = new Map();

function addSeen(hash) { seen.set(hash, Date.now()); }
function hasSeen(hash) {
	const ts = seen.get(hash);
	if (!ts) return false;
	if (Date.now() - ts > EXPIRY_MS) { seen.delete(hash); return false; }
	return true;
}
setInterval(() => {
	const now = Date.now();
	for (const [hash, ts] of seen) if (now - ts > EXPIRY_MS) seen.delete(hash);
}, EXPIRY_MS);

// --- MQTT clients ---
const inputClient = mqtt.connect(INPUT_BROKER, {
	...(INPUT_USER && { username: INPUT_USER, password: INPUT_PASS }),
	clientId: 'meshcore-decoder-in'
});

const outputClient = mqtt.connect(OUTPUT_BROKER, {
	username: OUTPUT_USER,
	password: OUTPUT_PASS,
	clientId: 'meshcore-decoder-out'
});

function publish(type, payload) {
	const topic = `meshcore/${type}`;
	outputClient.publish(topic, JSON.stringify(payload), { qos: 0 }, (err) => {
		if (err) process.stderr.write(`Publish error [${topic}]: ${err.message}\n`);
	});
}

inputClient.on('connect', () => {
	process.stderr.write(`Input connected: ${INPUT_BROKER}\n`);
	inputClient.subscribe('meshcore/#', (err) => {
		if (err) process.stderr.write(`Subscribe error: ${err}\n`);
		else process.stderr.write('Subscribed to meshcore/#\n');
	});
});

outputClient.on('connect', () => {
	process.stderr.write(`Output connected: ${OUTPUT_BROKER}\n`);
});

inputClient.on('message', (topic, message) => {
	if (!topic.endsWith('/packets')) return;
	try {
		const json = JSON.parse(message.toString());
		if (!json.raw || !json.hash) return;
		if (hasSeen(json.hash)) return;
		addSeen(json.hash);

		const packet = MeshCoreDecoder.decode(json.raw, { keyStore });
		if (!packet?.payload?.decoded) return;
		const { type } = packet.payload.decoded;

		const path = packet.path || [];
		const hops = path.length;
		const pathHashBytes = path.length > 0 ? path[0].length / 2 : null;

		const common = {
			receivedTimestamp: json.timestamp,
			receivedUnix: Math.floor(new Date(json.timestamp).getTime() / 1000),
			date: json.date,
			time: json.time,
			packetHash: json.hash,
			path,
			hops,
			pathHashBytes,
			snr: json.SNR,
			rssi: json.RSSI,
			score: json.score
		};

		const { type: _type, ...decodedRest } = packet.payload.decoded;
		const typeName = PAYLOAD_TYPES[type] || `unknown_${type}`;

		// Advert — enrich with structured fields
		if (type === 4) {
			const { name, location, deviceRole } = packet.payload.decoded.appData || {};
			const { publicKey, signatureValid } = packet.payload.decoded;
			publish('advert', {
				...common,
				...(publicKey !== undefined && { publicKey }),
				...(signatureValid !== undefined && { signatureValid }),
				...(name && { name }),
				...(location && { lat: location.latitude, lon: location.longitude }),
				...(deviceRole !== undefined && {
					deviceRole,
					deviceRoleName: DEVICE_ROLES[deviceRole] || `unknown(${deviceRole})`
				}),
				decoded: decodedRest
			});
			return;
		}

		// Group text — enrich with channel and message fields
		if (type === 5) {
			const channelHash = packet.payload.decoded.channelHash?.toUpperCase();
			const channel = channelNames[channelHash] || `unknown(${channelHash})`;
			const { sender, message: msg, timestamp: sentUnix } = packet.payload.decoded.decrypted || {};
			const receivedUnix = common.receivedUnix;
			publish('message', {
				...common,
				channel,
				channelHash,
				sentUnix: sentUnix || null,
				propagationSecs: sentUnix ? receivedUnix - sentUnix : null,
				...(sender !== undefined && { sender }),
				...(msg !== undefined && { message: msg }),
				decoded: decodedRest
			});
			return;
		}

		// Everything else — publish raw decoded payload under its type name
		publish(typeName, { ...common, decoded: decodedRest });
	} catch (e) {
		process.stderr.write(`Decode error: ${e.message}\n`);
	}
});

inputClient.on('error', (err) => process.stderr.write(`Input MQTT error: ${err.message}\n`));
outputClient.on('error', (err) => process.stderr.write(`Output MQTT error: ${err.message}\n`));
